packages/@react-stately/layout/src/TableLayout.ts

```diff
@@ -431,79 +491,125 @@ export class TableLayout<T> extends ListLayout<T> {
         }
         break;
       }
-      case 'rowgroup': {
-        let firstVisibleRow = this.binarySearch(node.children, rect.topLeft, 'y');
-        let lastVisibleRow = this.binarySearch(node.children, rect.bottomRight, 'y');
+      case 'rowgroup':
+      case 'section': {
+        if (node.layoutInfo.type === 'section') {
+          let sectionHeader = node.children[0];
+          // Add the section header to the visible layout infos since we always need to render it for labelling.
+          res.push(sectionHeader.layoutInfo);
+          for (let child of sectionHeader.children) {
+            res.push(child.layoutInfo);
+          }
+        }
 
-        // Add persisted rows before the visible rows.
-        let persistedRowIndices = this.persistedIndices.get(node.layoutInfo.key);
+        let persistedChildren = this.persistedIndices.get(node.layoutInfo.key);
         let persistIndex = 0;
-        while (
-          persistedRowIndices &&
-          persistIndex < persistedRowIndices.length &&
-          persistedRowIndices[persistIndex] < firstVisibleRow
-        ) {
-          let idx = persistedRowIndices[persistIndex];
-          if (idx < node.children.length) {
+
+        // If the children don't overlap the visible rect, just add all persisted nodes to the visible layout
+        let childrenRectTop = node.children[0].layoutInfo.rect.y;
+        let childrenRectBottom = node.children[node.children.length - 1].layoutInfo.rect.maxY;
+        if (persistedChildren && (childrenRectTop > rect.bottomRight.y || childrenRectBottom < rect.topLeft.y)) {
+          for (let idx of persistedChildren) {
             res.push(node.children[idx].layoutInfo);
             this.addVisibleLayoutInfos(res, node.children[idx], rect);
           }
-          persistIndex++;
-        }
-
-        for (let i = firstVisibleRow; i <= lastVisibleRow; i++) {
-          // Skip persisted rows that overlap with visible cells.
-          while (persistedRowIndices && persistIndex < persistedRowIndices.length && persistedRowIndices[persistIndex] < i) {
+        } else {
+          // Note:
```

Comment:
A variety of notes here to remind myself of the format of various pieces of data, happy to remove if people feel like it isn't helpful